File: algorithms/coding/hammingCode.cpp
#include "string"
#include "cmath"
#include "sstream"

uint16_t get_parity_bits_count(size_t msg_len) {
    if (msg_len == 1) {
        return 1;
    }

    uint16_t p = static_cast<uint16_t>(log2(msg_len + 1));

    return p; 
}

std::string hammingCodeEncode(const std::string &s) {
    if (s.emplty()) {
        return ""
    }
    
    uint16_t parity_bits_count = std::log2(s.size()) + 1;

    size_t str_ptr = 0;
    uint64_t parity_bits_val = 0;

    for (size_t i = 1; i < s.size() + 1 + parity_bits_count; ++i) {
        if ((i & (i - 1)) != 0 && s[str_ptr++] == '1') {  // if not power of 2 and s[i] == '1'
            parity_bits_val ^= i;
        }
    }

    std::stringstream ss;
    str_ptr = 0;
    for (size_t i = 1; i < s.size() + 1 + parity_bits_count; ++i) {
        if ((i & (i - 1)) == 0) {  // if power of 2
            ss << (((parity_bits_val & i) == i) ? '1' : '0');
        } else {
            ss << s[str_ptr++];
        }
    }

    return ss.str();
}

std::string hammingCodeDecode(const std::string &s) {
    std::stringstream ss;
    uint64_t parity_bits_val = 0;

    for (size_t i = 1; i < s.size() + 1; ++i) {
        if (s[i - 1] == '1') {
            parity_bits_val ^= i;
        }
    }

    for (size_t i = 1; i < s.size() + 1; ++i) {
        if ((i & (i - 1)) != 0) {  // if not power of 2
            char ch = s[i - 1];
            if (parity_bits_val == i) {  // if error occurred on this index
                ch = (ch == '1' ? '0' : '1');
            }
            ss << ch;
        }
    }

    return ss.str();
}

File: algorithms/coding/hammingCode.py
from math import log2, ceil


def get_parity_bits_count(msg_len: int) -> int:
    if msg_len == 1:
        return 1

    p = int(log2(msg_len + 1))

    while 2**p < msg_len + p + 1:
        p += 1

    return p


def hamming_code_encode(s: str) -> str:
    if not s:
        return ""

    parity_bits_count = get_parity_bits_count(len(s))
    str_ptr = 0
    parity_bits_val = 0

    for i in range(1, len(s) + 1 + parity_bits_count):
        if (i & (i - 1)) != 0:
            if s[str_ptr] == "1":  # if not power of 2 and s[i] == '1'
                parity_bits_val ^= i
            str_ptr += 1

    res: list[str] = []
    str_ptr = 0
    for i in range(1, len(s) + 1 + parity_bits_count):
        if (i & (i - 1)) == 0:  # if power of 2
            res.append("1" if ((parity_bits_val & i) == i) else "0")
        else:
            res.append(s[str_ptr])
            str_ptr += 1

    return "".join(res)


def hamming_code_decode(s: str) -> str:
    if len(s) == 2:
        if s[0] == s[1]:
            return s[1]
        else:
            return "0" if (s[1] == "1") else "1"

    res: list[str] = []
    parity_bits_val = 0

    for i in range(1, len(s) + 1):
        if s[i - 1] == "1":
            parity_bits_val ^= i

    for i in range(1, len(s) + 1):
        if (i & (i - 1)) != 0:  # if not power of 2
            ch = s[i - 1]
            if parity_bits_val == i:  # if error occurred on this index
                ch = "0" if (ch == "1") else "1"
            res.append(ch)
    return "".join(res)
